neuralmonkey/decoding_function.py

```diff
@@ -1,59 +1,93 @@
+        initial_state = tf.reshape(tf.tile(initial_state,
+                                           tf.shape(decoder_inputs[0])[:1]),
+                                   [-1, state_size])
 
     with tf.variable_scope(scope or "attention_decoder"):
-        batch_size = tf.shape(decoder_inputs[0])[0]    # Needed for reshaping.
-
-        # do manualy broadcasting of the initial state if we want it
-        # to be the same for all inputs
-        if len(initial_state.get_shape()) == 1:
-            state_size = initial_state.get_shape()[0].value
-            initial_state = tf.reshape(tf.tile(initial_state,
-                                               tf.shape(decoder_inputs[0])[:1]),
-                                       [-1, state_size])
-
-        state = initial_state
-        outputs = []
-        prev = None
-
-        attns = [a.initialize(batch_size, dtype) for a in attention_objects]
-
-        states = []
-        for i, inp in enumerate(decoder_inputs):
-            if i > 0:
-                tf.get_variable_scope().reuse_variables()
-            # If loop_function is set, we use it instead of decoder_inputs.
-            if loop_function is not None and prev is not None:
-                with tf.variable_scope("loop_function", reuse=True):
-                    inp = loop_function(prev, i)
-            # Merge input and previous attentions into one vector of the right
-            # size.
-            x = tf.nn.seq2seq.linear([inp] + attns, embedding_size, True)
-            # Run the RNN.
-            cell_output, state = cell(x, state)
-            states.append(state)
-            # Run the attention mechanism.
-            attns = [a.attention(state) for a in attention_objects]
 
-            if attns:
-                with tf.variable_scope("AttnOutputProjection"):
-                    output = tf.nn.seq2seq.linear([cell_output] + attns, output_size,
-                                             True)
+        output, state = decode_step(decoder_inputs[0], initial_state,
+                                    attention_objects, cell,
+                                    maxout_size, output_size)
+        outputs.append(output)
+        states.append(state)
+
+        for step in range(1, len(decoder_inputs)):
+            tf.get_variable_scope().reuse_variables()
+
+            if loop_function:
+                decoder_input = loop_function(state, step)
             else:
-                output = cell_output
+                decoder_input = decoder_inputs[step]
 
-            if loop_function is not None:
-                prev = output
+            output, state = decode_step(decoder_input, state, attention_objects,
+                                        cell, maxout_size, output_size)
             outputs.append(output)
+            states.append(state)
 
     return outputs, states
 
+
+def decode_step(prev_output, prev_state, attention_objects, rnn_cell,
+                maxout_size, output_size):
+    ## compute c_i:
```

Comment:
Jenže oni to nemaj očíslovaný.. :-) Jedině napsat do docstringu že to implementuje rovnice ze sekce A.2.2 na stranách 13 a 14..